### PR TITLE
8284619: javax/swing/JTable/8236907/LastVisibleRow.java failed with "Test Case Failed!!, Last Row not Visible!!!"

### PR DESCRIPTION
**Pull request by TejeshR13**
Fix history - #7422.
The test is automatic and check if the last row is painted (Which was not happening before the fix, the last row appeared only after MouseClick operation) by capturing the last row before and after MouseClick Operation.If the both the captured images are same, then the test passes else it fails.
This Fix includes for ubuntu and macos failure, in which only the test case failed due to following reasons.

In ubuntu, the test fails due to unexpected dark line appearing on right side of last row before MouseClick operation.
In macos, the test fails since after MouseClick operation of last row, clearSelection didn't show any effect on the selected row.
The issue is intermittent and not able to reproduce it frequently. As a defensive fix, the rectangular area which is being captured is reduce for ubuntu failure and Delay is added between ClearSelection of focus and Image Capture for macos failure

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.java.net/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8284619](https://bugs.openjdk.org/browse/JDK-8284619): javax/swing/JTable/8236907/LastVisibleRow.java failed with "Test Case Failed!!, Last Row not Visible!!!"

### Reviewers
 * [Prasanta Sadhukhan](https://openjdk.java.net/census#psadhukhan) (@prsadhuk - **Reviewer**) ⚠️ Review applies to [8722b1b0](https://git.openjdk.org/jdk/pull/9117/files/8722b1b020f30d54e2eab2122c2b1777a6e5182a)
 * [Harshitha Onkar](https://openjdk.java.net/census#honkar) (@honkar-jdk - Author)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/9117/head:pull/9117` \
`$ git checkout pull/9117`

Update a local copy of the PR: \
`$ git checkout pull/9117` \
`$ git pull https://git.openjdk.org/jdk pull/9117/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 9117`

View PR using the GUI difftool: \
`$ git pr show -t 9117`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/9117.diff">https://git.openjdk.org/jdk/pull/9117.diff</a>

</details>

### CODE REVIEW

**Review by prsadhuk — COMMENTED** [review]
Was there any reason of calling table.setFocusable(false) in clearSelect()? We have some focus related mach5 issues in macos that crops out intermittently so maybe we should get rid of that if it's not needed.
I think just calling clearSelection() would have been enough. Can you please check?

**Comment by TejeshR13** [comment]
> Was there any reason of calling table.setFocusable(false) in clearSelect()? We have some focus related mach5 issues in macos that crops out intermittently so maybe we should get rid of that if it's not needed. I think just calling clearSelection() would have been enough. Can you please check?

Yeah, table.setFocusable(false) is added intentionally, it is added to disable the cell focus. After selection the focus remains in cell, so in order to release the focus I used it. Any alternative to this is available?

**Comment by prsadhuk** [comment]
> > Was there any reason of calling table.setFocusable(false) in clearSelect()? We have some focus related mach5 issues in macos that crops out intermittently so maybe we should get rid of that if it's not needed. I think just calling clearSelection() would have been enough. Can you please check?
> 
> Yeah, table.setFocusable(false) is added intentionally, it is added to disable the cell focus. After selection the focus remains in cell, so in order to release the focus I used it. Any alternative to this is available?

Even if the focus remains in the cell, what was the problem? I guess the problem was the contents was not drawn initially and only was getting drawn if rowcell is clicked. So, before and mouseclick the bufferedimage contents should be same, I guess focus is irrelevant here.

**Comment by prsadhuk** [comment]
> > > > Was there any reason of calling table.setFocusable(false) in clearSelect()? We have some focus related mach5 issues in macos that crops out intermittently so maybe we should get rid of that if it's not needed. I think just calling clearSelection() would have been enough. Can you please check?
> > > 
> > > 
> > > Yeah, table.setFocusable(false) is added intentionally, it is added to disable the cell focus. After selection the focus remains in cell, so in order to release the focus I used it. Any alternative to this is available?
> > 
> > 
> > Even if the focus remains in the cell, what was the problem? I guess the problem was the contents was not drawn initially and only was getting drawn if rowcell is clicked. So, before and after mouseclick, the bufferedimage contents should be same, I guess focus is irrelevant here.
> 
> No issue with the contents drawn. Before mouseclick there wouldn't be any focus on any cell, but after mouseclick the focus will be seen on the clicked cell. In order to remove that set focus, setFocusable(false) is required. Otherwise the matching fails due to cell focus draw as rectangular border around the cell.

ok

**Review by prsadhuk — APPROVED** [review]

File: test/jdk/javax/swing/JTable/8236907/LastVisibleRow.java [hunk 1]
```diff
@@ -195,19 +169,13 @@ public void run() {
                 Point point = new Point(cellRect.x, cellRect.y);
                 SwingUtilities.convertPointToScreen(point, table);
 
-                captureRect[0] = new Rectangle(point.x, point.y, table.getColumnCount() * cellRect.width, cellRect.height);
+                captureRect[0] = new Rectangle(point.x+5, point.y+2, table.getColumnCount() * cellRect.width - 10, cellRect.height-2);
```
Comment (honkar-jdk) [inline test/jdk/javax/swing/JTable/8236907/LastVisibleRow.java:172]: Please adjust line lengths. Some of the lines contain more than 80 characters.

**Review by honkar-jdk — CHANGES_REQUESTED** [review]

**Comment by honkar-jdk** [comment]
In getCellClickPoint() method, you could simply the code by returning point variable , instead of using extra result variable. https://github.com/openjdk/jdk/blob/8722b1b020f30d54e2eab2122c2b1777a6e5182a/test/jdk/javax/swing/JTable/8236907/LastVisibleRow.java#L102

**Comment by honkar-jdk** [comment]
If the test criteria is to check if the last row is visible, then the following suggestion might help to make the test more stable on different platforms.

This is similar to one of the issues I was working on and @[azuev-java](https://github.com/azuev-java) suggested a good idea, that I think could be applicable here too.

- Set the background of last row (can be done for all the cells of the third row) to a different color, say red.
- Check if this color is visible using Robot.getPixelColor() at desired location.
- If red is returned then the third row is visible, else throw RuntimeException()

This would simplify the test case as you wouldn't have to compare two buffered images - before and after images and deal with platform specific issues such as unexpected dark line (on Ubuntu) and clearing the focus (on Mac). 

Hope this helps :)

File: test/jdk/javax/swing/JTable/8236907/LastVisibleRow.java (hunk 1 above)
Comment (TejeshR13) [reply test/jdk/javax/swing/JTable/8236907/LastVisibleRow.java:172]: Updated.

**Comment by TejeshR13** [comment]
> If the test criteria is to check if the last row is visible, then the following suggestion might help to make the test more stable on different platforms.
> 
> This is similar to one of the issues I'm working on and @[azuev-java](https://github.com/azuev-java) suggested a good idea, that I think could be applicable here too.
> 
> * Set the background of ONLY the last row to a different color, say red. (can be done for all the cells of the third row)
> * Check if this color is visible using Robot.getPixelColor() at desired location.
> * If red is returned then the third row is visible, else throw RuntimeException()
> 
> This would simplify the test case as you wouldn't have to compare two buffered images - before and after images and deal with platform specific issues such as unexpected dark line (on Ubuntu) and clearing the focus (on Mac).
> 
> Hope this helps :)

Its really helpful suggestion. But in this case we are validating the last row contents, not the last row itself. Before the fix the last row was painted without the contents, after the fix we are validating whether the contents are printed on it. We still can use the technique though to verify the contents, means have to get the text content position and check if it is black or still background color. But I feel for this test case this technique wont be robust enough and finding the text area inside the cell might not be accurate always.

**Comment by TejeshR13** [comment]
> In getCellClickPoint() method, you could simplify the code by returning point variable , instead of using extra result variable.
> 
> https://github.com/openjdk/jdk/blob/8722b1b020f30d54e2eab2122c2b1777a6e5182a/test/jdk/javax/swing/JTable/8236907/LastVisibleRow.java#L102

Updated.

**Review by honkar-jdk — APPROVED** [review]

**Pull request closed without merging**